### ISSUE TITLE
abspath strips one backslash from the Windows share path

### ISSUE THREAD

**Issue by jaakkor2** [JuliaLang/julia#18532]
On Windows 7, abspath() does not work correctly for a path of type `\\[server]\[sharename]`. The path becomes incorrectly `\[server]\[sharename]`.  As a consequence, I cannot `include` a file even though `readdir()` shows it exists.

The following example shows the incorrect behavior, replace `[server]`and `[sharename]`with proper values.

``` julia
cd(s"\\[server]\[sharename]")
cd(abspath(pwd())) # this does not work
ERROR: chdir \[server]\[sharename]: no such file or directory (ENOENT)
 in uv_error at .\libuv.jl:68 [inlined]
 in cd(::String) at .\file.jl:39
cd(pwd()) # this works
```

Tested versions that are affected are 0.4.6 and 0.5.0-rc4.

**Comment by vtjnash** [comment]
Should we avoid calling `normpath` in as many places? On NT, the result is an equivalent path (sans bugs), but on unix, it is not (for example, if a directory is a symlink or mount)

**Comment by jaakkor2** [comment]
If the server-string contains a hyphen, then one of the leading backslashes is wrongly stripped.  But if it does not, should a backslash added to the end of the path, maybe not?  And if the sharename-string contains a hyphen, looks like a backslash is weirdly added before the hyphen.
```
julia> abspath(s"\\aaa-bbb\ccc")
"\\aaa-bbb\\ccc"
julia> abspath(s"\\aaabbb\ccc")
"\\\\aaabbb\\ccc\\"
julia> abspath(s"\\aaa\bbb-ccc")
"\\\\aaa\\bbb\\-ccc"
```

**Comment by jaakkor2** [comment]
The culprit seems to be in file `base/path.jl` the [line 38](https://github.com/JuliaLang/julia/blob/a304c557928d9eaa0a28af5fe96327d5da70d14b/base/path.jl#L38)
```julia
m = match(r"^(\w+:|\\\\\w+\\\w+|\\\\\?\\UNC\\\w+\\\w+|\\\\\?\\\w+:|)(.*)$", path)
```

The regex word character `\w`neglects some allowed characters. Windows UNC allowed characters described here
https://msdn.microsoft.com/en-us/library/gg465305.aspx

The relevance of this bug is that I cannot run a file from a Windows network share, when the sharename containt a hyphen in my case, or some of the other allowed characters not captured by the `\w`.

**Comment by tkelman** [comment]
would https://github.com/JuliaLang/julia/pull/19695 fix this then? that pr could use a couple more tests and a little bit of cleanup but otherwise seemed okay

**Comment by jaakkor2** [comment]
Yes, since those test fail for abspath as well. `.` is not captured by `\w` either.
```julia
julia> abspath(s"\\servername\hello.world\filename.ext")
"\\\\servername\\hello\\.world\\filename.ext"

julia> abspath(s"\\servername.com\hello.world\filename.ext")
"\\servername.com\\hello.world\\filename.ext"
```

**Comment by StefanKarpinski** [comment]
I'm still unclear on what the fix here is...

**Comment by tkelman** [comment]
fixed by #19695 I believe
